src/Components/src/Microsoft.AspNetCore.Components.Server/Rendering/HtmlRenderer.cs

```diff
@@ -0,0 +1,207 @@
+        /// <returns>A sequence of <see cref="string"/> fragments that represent the HTML text of the component.</returns>
+        private IEnumerable<string> RenderComponent(Type componentType, ParameterCollection initialParameters)
+        {
+            var tree = CreateInitialRender(componentType, initialParameters);
+
+            var frames = tree.GetFrames();
+
+            if (frames.Count == 0)
+            {
+                return Array.Empty<string>();
+            }
+            else
+            {
+                var (segments, newPosition) = RenderFrames(frames, 0, frames.Count);
+                Debug.Assert(newPosition == frames.Count);
+                return segments;
+            }
+        }
+
+        private (IEnumerable<string>, int newPosition) RenderFrames(ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+        {
+            IEnumerable<string> segments = Array.Empty<string>();
+            IEnumerable<string> nextSegments;
+            var nextPosition = position;
+            var endPosition = position + maxElements;
+            while (position < endPosition)
+            {
+                (nextSegments, nextPosition) = RenderCore(frames, position, maxElements);
+                if (position == nextPosition)
+                {
+                    throw new InvalidOperationException("We didn't consume any input.");
+                }
+                position = nextPosition;
+                segments = segments.Concat(nextSegments);
+            }
+
+            return (segments, nextPosition);
+        }
+
+        private (IEnumerable<string> segments, int nextPosition) RenderCore(
+            ArrayRange<RenderTreeFrame> frames,
+            int position,
+            int length)
+        {
+            var frame = frames.Array[position];
+            switch (frame.FrameType)
+            {
+                case RenderTreeFrameType.Element:
+                    return RenderElement(frames, position);
+                case RenderTreeFrameType.Attribute:
+                    return RenderAttributes(frames, position, 1);
+                case RenderTreeFrameType.Text:
+                    return (new[] { frame.TextContent }, ++position);
+                case RenderTreeFrameType.Markup:
+                    return (new[] { frame.MarkupContent }, ++position);
+                case RenderTreeFrameType.Component:
+                    return RenderChildComponent(frames, position);
+                case RenderTreeFrameType.Region:
+                    return RenderFrames(frames, position + 1, frame.RegionSubtreeLength - 1);
+                case RenderTreeFrameType.ElementReferenceCapture:
+                case RenderTreeFrameType.ComponentReferenceCapture:
+                    return (Array.Empty<string>(), ++position);
+                default:
+                    throw new InvalidOperationException("Invalid element frame type.");
```

Comment:
Nit: Might be handy if the exception message specifies what the invalid frame type was, in case we add more frame types later.